### PR TITLE
Block imports of files via a whitelist

### PR DESCRIPTION
**Pull request by rk-for-zulip**
Add and use a custom ESLint plugin to prevent certain files from being imported unless they're on a whitelist.

This also sets up a minimal framework for other custom ESLint rules.

### CODE REVIEW

**Comment by ray-kraesig** [comment]
So this is me coming back from the zoo with an armful of yak hair. 

Ultimately, this is in support of #3881. From the top (bottom?):

* #3881 involves modifying `outboxReducer`. New tests are needed.
* `outboxReducer-tests` isn't properly typed yet. In fact there are bugs in the test code because of that.
* Writing well-typed testing code for a reducer is an absolutely *miserable* experience if you don't have access to the types of the actions it's reducing. (Particularly server-originating actions.)
* Just adding `export` to the action types would result in those types leaking into places where `Action` should be opaque. (See, e.g., commit 6b703b6dd28f2c09f0d65db3facf86f7f9166e3c.)
* Moving them to a new conditionally-import-restricted file is possible, but makes the `overrides` property in `.eslintrc` much more complicated to write and test.
  * There are two available ESLint rules that look like they *should* be able to do this, but neither one actually can. (`import/no-restricted-paths` is particularly confusingly documented.)

And at that point I'd spent enough time digging around in the guts of ESLint that I figured it would be easy enough to adapt one of those rules into a custom rule, and, well, here we are.

This isn't even the end of the subtask chain – as that first commit indicates, that's actually #3969 (which is also needed for, or at least badly wanted by, the translation automation tool).

**Comment by ray-kraesig** [comment]
Updates (if you've looked at this before):
* Moved the free-floating `arrayOfStrings` inherited from `no-restricted-imports` into the JSON schema as a definition.
* Removed `importerBlacklist`, since synthesizing it from `importerWhitelist` is relatively easy.

File: .eslintplugin/zulip-disallow-imports.js [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+/**
+ * Basic set of error messages.
+ *
+ * These error messages also have automatically-generated versions with a
+ * trailing `{{customMessage}}`; see `messages`, below.
+ */
+const baseMessages = {
+  fileImport: "'{{importee}}' may not be imported.",
+  fileWhitelist: "'{{importer}}' may not import '{{importee}}' (not whitelisted).",
+  fileBlacklist: "'{{importer}}' may not import '{{importee}}' (blacklisted).",
```
Comment (ray-kraesig) [inline .eslintplugin/zulip-disallow-imports.js]: (Note to self: remove this line.)

**Comment by gnprice** [comment]
This would be a useful lint feature, but not important enough to match the complexity cost of the implementation, when that's amortized only over our small codebase.

I think it would make a lot of sense for this feature to exist in ESLint upstream, or in an ESLint plugin advertised to the broader JS community. Also for any one codebase with maybe a few hundred people working in it, I think it'd be worthwhile for someone working on code quality there to build something like this.

**Pull request closed without merging**